### ISSUE TITLE
BERT Large/TF 1.15:  NaN on multinode horovod  pre-training

### ISSUE THREAD

**Issue by whatdhack** [NVIDIA/DeepLearningExamples#549]
Related to **Model/Framework(s)** 
BERT Large /TF 1.15 pre-training 

**Describe the bug**
Keep seeing NaN on multi-node pre-training. 

```
Skipping time record for  1  due to checkpoint-saving/warmup overhead
INFO:tensorflow:loss = 11.222988, step = 1
I0606 09:26:10.654364 46912496659904 basic_session_run_hooks.py:262] loss = 11.222988, step = 1
INFO:tensorflow:loss = 11.166957, step = 1
I0606 09:26:13.364254 46912496659904 basic_session_run_hooks.py:262] loss = 11.166957, step = 1
INFO:tensorflow:loss = 11.153452, step = 1
I0606 09:26:13.499738 46912496659904 basic_session_run_hooks.py:262] loss = 11.153452, step = 1
INFO:tensorflow:loss = 11.164111, step = 1
I0606 09:26:13.518184 46912496659904 basic_session_run_hooks.py:262] loss = 11.164111, step = 1
ERROR:tensorflow:Model diverged with loss = NaN.
E0606 09:34:26.659411 46912496659904 basic_session_run_hooks.py:760] Model diverged with loss = NaN.
Skipping time record for  2  due to checkpoint-saving/warmup overhead
ERROR:tensorflow:Model diverged with loss = NaN.
E0606 09:34:26.717127 46912496659904 basic_session_run_hooks.py:760] Model diverged with loss = NaN.
```
**To Reproduce**
Using all stock software. 

**Expected behavior**
Pre-training proceeding without NaN

**Environment**
Please provide at least:
* Container version (e.g. pytorch:19.05-py3): Bare metal 
* GPUs in the system: (e.g. 8x Tesla V100-SXM2-16GB):1xV100 per node
* CUDA driver version (e.g. 418.67):440.

Using the following script.

```
TensorFlow/LanguageModeling/BERT/scripts/run_pretraining_lamb_phase1.sh
```

Here are my hyper-parameters on a 4 node Horovod cluster. 

```
0606 09:07:33.429131 46912496659904 run_pretraining.py:582]   dllog_path: /tmp/bert_dllog.json
INFO:tensorflow:***** Running training *****
I0606 09:07:33.429270 46912496659904 run_pretraining.py:624] ***** Running training *****
INFO:tensorflow:  init_checkpoint: None
I0606 09:07:33.429659 46912496659904 run_pretraining.py:582]   init_checkpoint: None
INFO:tensorflow:  optimizer_type: lamb
I0606 09:07:33.430222 46912496659904 run_pretraining.py:582]   optimizer_type: lamb
INFO:tensorflow:  Batch size = 8
I0606 09:07:33.426508 46912496659904 run_pretraining.py:625]   Batch size = 8
INFO:tensorflow:  max_seq_length: 512
I0606 09:07:33.430755 46912496659904 run_pretraining.py:582]   max_seq_length: 512
INFO:tensorflow:  max_predictions_per_seq: 76
I0606 09:07:33.431279 46912496659904 run_pretraining.py:582]   max_predictions_per_seq: 76
INFO:tensorflow:  Batch size = 8
INFO:tensorflow:  do_train: True
I0606 09:07:33.431803 46912496659904 run_pretraining.py:582]   do_train: True
I0606 09:07:33.428258 46912496659904 run_pretraining.py:625]   Batch size = 8
INFO:tensorflow:  do_eval: False
I0606 09:07:33.432330 46912496659904 run_pretraining.py:582]   do_eval: False
INFO:tensorflow:  train_batch_size: 8
I0606 09:07:33.432853 46912496659904 run_pretraining.py:582]   train_batch_size: 8
INFO:tensorflow:  Batch size = 8
INFO:tensorflow:  eval_batch_size: 8
I0606 09:07:33.433381 46912496659904 run_pretraining.py:582]   eval_batch_size: 8
I0606 09:07:33.430223 46912496659904 run_pretraining.py:625]   Batch size = 8
INFO:tensorflow:  learning_rate: 0.0001
I0606 09:07:33.433963 46912496659904 run_pretraining.py:582]   learning_rate: 0.0001
INFO:tensorflow:  num_train_steps: 10000
I0606 09:07:33.434523 46912496659904 run_pretraining.py:582]   num_train_steps: 10000
INFO:tensorflow:  num_warmup_steps: 0
I0606 09:07:33.435056 46912496659904 run_pretraining.py:582]   num_warmup_steps: 0
INFO:tensorflow:  save_checkpoints_steps: 2000
I0606 09:07:33.435584 46912496659904 run_pretraining.py:582]   save_checkpoints_steps: 2000
INFO:tensorflow:  display_loss_steps: 10
I0606 09:07:33.436110 46912496659904 run_pretraining.py:582]   display_loss_steps: 10
INFO:tensorflow:  iterations_per_loop: 1000
I0606 09:07:33.436651 46912496659904 run_pretraining.py:582]   iterations_per_loop: 1000
INFO:tensorflow:  max_eval_steps: 100
I0606 09:07:33.437182 46912496659904 run_pretraining.py:582]   max_eval_steps: 100
INFO:tensorflow:  num_accumulation_steps: 1
I0606 09:07:33.437711 46912496659904 run_pretraining.py:582]   num_accumulation_steps: 1
INFO:tensorflow:  allreduce_post_accumulation: False
I0606 09:07:33.438279 46912496659904 run_pretraining.py:582]   allreduce_post_accumulation: False
INFO:tensorflow:  verbose_logging: False
I0606 09:07:33.438812 46912496659904 run_pretraining.py:582]   verbose_logging: False
INFO:tensorflow:  horovod: True
I0606 09:07:33.439346 46912496659904 run_pretraining.py:582]   horovod: True
INFO:tensorflow:  report_loss: True
I0606 09:07:33.439877 46912496659904 run_pretraining.py:582]   report_loss: True
INFO:tensorflow:  manual_fp16: False
I0606 09:07:33.440401 46912496659904 run_pretraining.py:582]   manual_fp16: False
INFO:tensorflow:  use_xla: False
I0606 09:07:33.440928 46912496659904 run_pretraining.py:582]   use_xla: False
INFO:tensorflow:  use_fp16: False
I0606 09:07:33.441452 46912496659904 run_pretraining.py:582]   use_fp16: False
INFO:tensorflow:  init_loss_scale: 4294967296
I0606 09:07:33.442025 46912496659904 run_pretraining.py:582]   init_loss_scale: 4294967296
INFO:tensorflow:  ?: False
I0606 09:07:33.442563 46912496659904 run_pretraining.py:582]   ?: False
INFO:tensorflow:  help: False
I0606 09:07:33.443088 46912496659904 run_pretraining.py:582]   help: False
INFO:tensorflow:  helpshort: False
I0606 09:07:33.443617 46912496659904 run_pretraining.py:582]   helpshort: False
INFO:tensorflow:  helpfull: False
I0606 09:07:33.444145 46912496659904 run_pretraining.py:582]   helpfull: False
INFO:tensorflow:  helpxml: False
I0606 09:07:33.444665 46912496659904 run_pretraining.py:582]   helpxml: False
```

**Comment by guotong1988** [comment]
Similar question.

**Comment by swethmandava** [comment]
LR is set according to the recipe mentioned in the [LAMB paper](https://arxiv.org/pdf/1904.00962.pdf). When batch size is reduced significantly (64 in your experiment vs 64x1024 in the shared recipe), LR needs to be tuned by reducing it.

**Comment by swethmandava** [comment]
LAMB expects an LR = 6e-3 for phase 1 and 4e-3 for phase2. for 8 nodes and 8 gpus, FLAGS.learning_rate = (LR / (num_gpu * num_nodes)) for the same batch size. 
The advantage of lamb is allowing you to use higher batch sizes. If you're using bs/gpu=8, why not use adam? 

The global step not changing is due to NaN potentially caused by the loss scaler being too high. the optimizer automatically adjusts loss scale value till it is optimal. You should see it start training after these warnings.

**Comment by whatdhack** [comment]
Thanks @swethmandava . My pre-training is for phase 2 - hence sequence length of 512.   A few simple  questions. 

1. What is the maximum bs/gpu at 512 sequence length  a V100 should be able to handle ?   
2. So theoretically NaN  occurs at  all the GPUs as  Allreduce does the exact same update in each node.  Is that also observed in practice, or NaN occurs at one node that eventually propagates to all the nodes ?
3.  In the paper , https://arxiv.org/pdf/1904.00962.pdf,  Table 8 showed AdamW with 16K batch size and LR of 1e-4.  So if I just run 1 node with bs of 8, my learning rate should be 1e-4*8/16K , right ?

**Comment by whatdhack** [comment]
@swethmandava , wondering  if you have any thoughts on the  above. Where can I find more details on the tests (e.g.  53 min | DGX SuperPOD | 92 x DGX-2H | 1,472 ) mentioned in https://docs.nvidia.com/ngc/multi-node-bert-user-guide/index.html  ?

**Comment by swethmandava** [comment]
Can you enable --use_fp16 and --use_xla? You should be able to fit 2-4 BS/GPU if you're using FP16 for seq_len 512. 

We used the following (configs)[https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/configurations.yml]

**Comment by githubsgi** [comment]
Hi @swethmandava , thanks 2-4  seqs per batch per gpu makes sense - using 3.   Thought fp32 should be more forgiving then fp16 , right ? Also, I am only looking at fp32,  and using adam by the way. See following  log messages. 

```
Stalled ranks:
0: [HorovodBroadcast_bert_embeddings_LayerNorm_beta_0, HorovodBroadcast_bert_embeddings_LayerNorm_beta_adam_m_0, HorovodBroadcast_bert_embeddings_LayerNorm_beta_adam_v_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_adam_m_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_adam_v_0 ...]
[2020-06-29 00:34:39.900478: W horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
Stalled ranks:
0: [HorovodBroadcast_bert_embeddings_LayerNorm_beta_0, HorovodBroadcast_bert_embeddings_LayerNorm_beta_adam_m_0, HorovodBroadcast_bert_embeddings_LayerNorm_beta_adam_v_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_adam_m_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_adam_v_0 ...]
.
.
Stalled ranks:
6: [DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_attention_output_LayerNorm_batchnorm_mul_grad_tuple_control_dependency_1_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_attention_output_LayerNorm_batchnorm_sub_grad_tuple_control_dependency_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_attention_output_dense_BiasAdd_grad_tuple_control_dependency_1_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_attention_output_dense_MatMul_grad_tuple_control_dependency_1_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_intermediate_dense_BiasAdd_grad_tuple_control_dependency_1_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_intermediate_dense_MatMul_grad_tuple_control_dependency_1_0 ...]

```

Tried to find knobs for more debug info on these.  Any  pointers ?

**Comment by whatdhack** [comment]
Hi @swethmandava , thanks 2-4 seqs per batch per gpu makes sense - using 3. Thought fp32 should be more forgiving than fp16 , right ? Also, I am only looking at fp32, and using adam by the way. See following log messages.

```
Stalled ranks:
0: [HorovodBroadcast_bert_embeddings_LayerNorm_beta_0, HorovodBroadcast_bert_embeddings_LayerNorm_beta_adam_m_0, HorovodBroadcast_bert_embeddings_LayerNorm_beta_adam_v_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_adam_m_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_adam_v_0 ...]
[2020-06-29 00:34:39.900478: W horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.
Stalled ranks:
0: [HorovodBroadcast_bert_embeddings_LayerNorm_beta_0, HorovodBroadcast_bert_embeddings_LayerNorm_beta_adam_m_0, HorovodBroadcast_bert_embeddings_LayerNorm_beta_adam_v_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_adam_m_0, HorovodBroadcast_bert_embeddings_LayerNorm_gamma_adam_v_0 ...]
.
.
Stalled ranks:
6: [DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_attention_output_LayerNorm_batchnorm_mul_grad_tuple_control_dependency_1_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_attention_output_LayerNorm_batchnorm_sub_grad_tuple_control_dependency_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_attention_output_dense_BiasAdd_grad_tuple_control_dependency_1_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_attention_output_dense_MatMul_grad_tuple_control_dependency_1_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_intermediate_dense_BiasAdd_grad_tuple_control_dependency_1_0, DistributedAdamWeightDecayOptimizer_Allreduce/HorovodAllreduce_gradients_bert_encoder_layer_19_intermediate_dense_MatMul_grad_tuple_control_dependency_1_0 ...]
```

**Comment by swethmandava** [comment]
Can you provide a reproducible script/command? @whatdhack
